### PR TITLE
Fix undefined table component

### PR DESCRIPTION
**Pull request by embooglement**
There is a race condition where if a header cell is removed from the dom during resizing, `tableComponent` becomes undefined in `elementSizeDidChange` because the associated view has already been destroyed by the time the `afterRender` portion of the run loop happens. This throws an exception. This commit properly cancels the scheduled callback when the view is destroyed, preventing the exception.

### CODE REVIEW

**Comment by jordansmith42** [comment]
LGTM @cyril-sf any concerns?

File: tests/unit/header-cell-test.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  unit: true,
+  needs: [
+    'template:header-cell'
+  ]
+});
+
+test('The elementSizeDidChange scheduled task is correctly canceled when the element is destroyed', function(assert) {
+  assert.expect(3);
+
+  let headerCell = this.subject({
```
Comment (philpee2) [inline tests/unit/header-cell-test.js]: I think this can be `const`

File: tests/unit/header-cell-test.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    content: ColumnDefinition.create({}),
+    tableComponent: {
+      columnMode: 'standard'
+    }
+  });
+
+  Ember.run(function() {
```
Comment (philpee2) [inline tests/unit/header-cell-test.js]: One-liner, bro
`Ember.run(() => headerCell.append());`

File: tests/unit/header-cell-test.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    headerCell.append();
+  });
+
+  assert.ok(!headerCell._scheduledElementSizeDidChange,
```
Comment (philpee2) [inline tests/unit/header-cell-test.js]: Hm, doesn't seem great that you're testing something that isn't the public API of this view

File: tests/unit/header-cell-test.js (hunk 2 above)
Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: `this.append` or `this.render` ( I forgot which version of Ember QUnit is used ). No need to wrap it in a run loop, no?

File: tests/unit/header-cell-test.js (hunk 3 above)
Comment (embooglement) [inline tests/unit/header-cell-test.js]: I agree, but I don't know a better way to do it.

File: tests/unit/header-cell-test.js (hunk 2 above)
Comment (embooglement) [inline tests/unit/header-cell-test.js]: If you really want me to, I guess.

File: tests/unit/header-cell-test.js (hunk 1 above)
Comment (embooglement) [inline tests/unit/header-cell-test.js]: Good catch

**Comment by cyril-sf** [comment]
Does the test fail if you revert the fix?

File: tests/unit/header-cell-test.js (hunk 2 above)
Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: It's using the proper API of the library we use, so yes :smile:

File: tests/unit/header-cell-test.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  needs: [
+    'template:header-cell'
+  ]
+});
+
+test('The elementSizeDidChange scheduled task is correctly canceled when the element is destroyed', function(assert) {
+  assert.expect(3);
+
+  const headerCell = this.subject({
+    content: ColumnDefinition.create({}),
```
Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: can you mock the column definition the same way you do with the `tableComponent`?

File: tests/unit/header-cell-test.js (hunk 2 above)
Comment (embooglement) [inline tests/unit/header-cell-test.js]: `this.append` is only defined for tests using `moduleForComponent`, and that only works with actual components, not views. `append` has to be called on the view directly, and must be in a run loop.

Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: I forgot that this was still a view :disappointed:

File: tests/unit/header-cell-test.js (hunk 4 above)
Comment (embooglement) [inline tests/unit/header-cell-test.js]: What do you mean? I only mocked the stuff in `tableComponent` that was necessary for the test to work.

Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: Do you need to import `ColumnDefinition`?

Comment (embooglement) [inline tests/unit/header-cell-test.js]: Yes.

Comment (embooglement) [inline tests/unit/header-cell-test.js]: There's a bunch of initialization stuff it does that's needed, it can't just be a raw object.

**Comment by embooglement** [comment]
Yes, things fail if the fix is changed incorrectly (the test doesn't even make sense if the fix is reverted because then `_scheduledElementSizeDidChange` is never even mentioned, so them failing is meaningless).

**Comment by cyril-sf** [comment]
That's right. I guess not canceling `_scheduledElementSizeDidChange` should be enough to make the test fail.

**Comment by embooglement** [comment]
Yeah, that fails.

**Comment by cyril-sf** [comment]
I realize that `_scheduledElementSizeDidChange` is canceled twice, it feels like we're missing a test now.

**Comment by embooglement** [comment]
How is it cancelled twice?

**Comment by embooglement** [comment]
Oh, like if the function is called without the element being destroyed.

**Comment by cyril-sf** [comment]
It's canceled in `elementSizeDidChange` and in `willDestroyElement`.

**Comment by cyril-sf** [comment]
By looking at the code, canceling in `willDestroyElement` seems actually not needed as it will be done in `elementSizeDidChange` anyway.

**Comment by embooglement** [comment]
I guess I code just always call `elementSizeDidChange` in `willDestroyElement`.

**Comment by embooglement** [comment]
@cyril-sf Added another test and updated `willDestroyElement`. Any other concerns?

**Comment by embooglement** [comment]
@Addepar/fire updated this with comments from @cyril-sf, any other concerns?

File: tests/unit/header-cell-test.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  Ember.run(function() {
+    headerCell.set('width', 100);
+    headerCell.destroy();
+  });
+
+  assert.equal(callCount, 1,
+    'elementSizeDidChange is not called for destroyed elements');
+});
+
+test('The elementSizeDidChange scheduled task is canceled when the task is done', function(assert) {
```
Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: `is called only once` instead of `is canceled`?

File: tests/unit/header-cell-test.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    oldElementSizeDidChange.apply(this, arguments);
+  };
+
+  Ember.run(() => headerCell.append());
+
+  assert.equal(callCount, 1,
+    'elementSizeDidChange is called for when initializing elements');
+
+  Ember.run(function() {
+    headerCell.set('width', 100);
```
Comment (cyril-sf) [inline tests/unit/header-cell-test.js]: Set the width twice to make sure that `scheduleOnce` is used

Comment (embooglement) [inline tests/unit/header-cell-test.js]: Done

**Comment by embooglement** [comment]
@cyril-sf anything else?

**Comment by cyril-sf** [comment]
LGTM

**Pull request merged**